controllers/k8ssandracluster_controller.go

```diff
@@ -215,23 +220,64 @@ func (r *K8ssandraClusterReconciler) reconcile(ctx context.Context, kc *api.K8ss
 				return ctrl.Result{}, err
 			}
 		}
+	}
 
-		if kc.HasStargates() && i == 0 {
-			// only needs to be done once per cluster, so we do it when reconciling the first dc only
-			if err := r.ensureStargateAuthKeyspaceExists(ctx, kc, actualDc, remoteClient, logger); err != nil {
-				return ctrl.Result{RequeueAfter: longDelay}, err
-			}
+	for _, dcTemplate := range kc.Spec.Cassandra.Datacenters {
+		dcConfig := cassandra.Coalesce(kc.Spec.Cassandra.DeepCopy(), dcTemplate.DeepCopy())
+		dcConfig.AdditionalSeeds = seeds
+		desiredDc, err := cassandra.NewDatacenter(kcKey, dcConfig)
+		dcKey := types.NamespacedName{Namespace: desiredDc.Namespace, Name: desiredDc.Name}
+		logger := kcLogger.WithValues("CassandraDatacenter", dcKey)
+
+		if err != nil {
+			logger.Error(err, "Failed to create new CassandraDatacenter")
+			return ctrl.Result{}, err
 		}
-		result, err := r.reconcileStargate(ctx, kc, dcTemplate, actualDc, logger, remoteClient)
-		if !result.IsZero() || err != nil {
-			return result, err
+
+		desiredDcHash := utils.DeepHashString(desiredDc)
+		desiredDc.Annotations[api.ResourceHashAnnotation] = desiredDcHash
+
+		actualDc := &cassdcapi.CassandraDatacenter{}
+
+		remoteClient, err := r.buildRemoteClient(logger, dcTemplate, ctx, dcKey, actualDc, kc)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		if err = remoteClient.Get(ctx, dcKey, actualDc); err == nil {
+			result, err := r.reconcileStargate(ctx, kc, dcTemplate, actualDc, logger, remoteClient)
+			if !result.IsZero() || err != nil {
+				return result, err
+			}
 		}
 	}
 
 	kcLogger.Info("Finished reconciling the k8ssandracluster")
 	return ctrl.Result{}, nil
 }
 
+func (r *K8ssandraClusterReconciler) buildRemoteClient(
+	logger logr.Logger,
+	dcTemplate api.CassandraDatacenterTemplate,
+	ctx context.Context,
+	dcKey types.NamespacedName,
+	actualDc *cassdcapi.CassandraDatacenter,
+	kc *api.K8ssandraCluster,
+) (client.Client, error) {
+	remoteClient, err := r.ClientCache.GetRemoteClient(dcTemplate.K8sContext)
```

Comment:
Even better: I think we don't need the `buildRemoteClient` method at all: there are safety checks already in `r.ClientCache.GetRemoteClient`, and the custom error message created line 269 doesn't really bring any new information compared to the original error message.

Reply:
I'm all for less code 🙂 
I'll remove the function and use `r.ClientCache.GetRemoteClient` only in both loops.